### ISSUE TITLE
Possible Bug: while_loop, map_fn do not parallize 

### ISSUE THREAD

**Issue by georgh** [tensorflow/tensorflow#12937]
The `parallel_iterations` parameter of `while_loop` and `map_fn` do not reduce the runtime as expected.
In comparison to the same operations created in a python loop, `while_loop` and `map_fn` are at least 4 times slower. `while_loop` does not scale at all

More details:
For my project, I need to calculate a matrix blockwise. 
I execute the calculation in a loop, each loop iteration working on a small part of the matrix, not accessing the rest of the matrix.
To achieve maximum performance, the loop has to be executed in parallel.
In addition, I need to control the number of parallel threads because of memory limitations.
I tried to use `parallel_iterations`, but changing this values doesn't change the runtime at all.

Using a python-for loop to create the same number of operations and syncing them with `control_dependencies` is around 4 times faster.
In contrast to `while_loop`, `map_fn` is getting faster with bigger `parallel_iterations` but remains noticeable slower then the python creation.
If this issue is known (didn't find anything at stackoverflow or here) and not resolvable, 
the documentation of while_loop needs to be improved heavily - to avoid this operation if the loop should be parallized.

Benchmark-Code:
```
import tensorflow as tf
import time

if __name__ == '__main__':
      runs = 100
      N = 100
      M = 30

      KernelRow = tf.Variable(tf.zeros([M, N*M]), name='KernelRow')

      def calEntry(KernelVariable, pos):
            op = tf.assign(KernelVariable[:, pos*M:(pos+1)*M], 
                  tf.eye(M) * 123. + tf.exp(tf.random_uniform([M, M]))) #perform some operations 
            with tf.control_dependencies([op]):
                  return tf.constant(0)

      def calRow(KernelVariable, N, parallelNum):
            ops = []
            for ic in range(0, N, parallelNum):
                  with tf.control_dependencies(ops):
                        for j in range(parallelNum):
                              if ic + j < N:
                                    ops += [calEntry(KernelVariable, ic+j)]
                        
            with tf.control_dependencies(ops):
                  return tf.identity(KernelVariable)

      def calRow_while(KernelVariable, N, parallelNum):
            i = tf.constant(0)
            condition = lambda i: tf.less(i, N)
            def body(ic):
                  updateKernel = calEntry(KernelVariable, ic)
                  with tf.control_dependencies([updateKernel]):
                        return ic + 1
            return tf.while_loop(condition, body, [i], back_prop=False, parallel_iterations=parallelNum)
            
      def calRow_map(KernelVariable, N, parallelNum):
            f = lambda x: calEntry(KernelVariable, x)
            return tf.map_fn(f, tf.range(N, dtype=tf.int32), parallel_iterations=parallelNum)

      paraOps = []
      paraOpsWhile = []
      paraOpsMap = []
      for paraEntry in [1, 10, 50, 100]:
            paraOps += [(calRow(KernelRow, N, paraEntry), paraEntry )]
            paraOpsWhile += [(calRow_while(KernelRow, N, paraEntry), paraEntry )]
            paraOpsMap += [(calRow_map(KernelRow, N, paraEntry), paraEntry )]

      init = tf.global_variables_initializer()
      #perform calculation
      with tf.Session() as sess:
            sess.run(init)
            for op in paraOps:
                  runtime = []
                  sess.run(op[0]) #warm up
                  for i in range(runs):
                        startTime = time.time()
                        sess.run(op[0])
                        runtime += [time.time() - startTime]
                  print("Calculation using {} threads took {:.4f} +- {:.4f}".format(op[1], np.mean(runtime), np.std(runtime) ))
            
            for op in paraOpsWhile:
                  runtime = []
                  sess.run(op[0]) #warm up
                  for i in range(runs):
                        startTime = time.time()
                        sess.run(op[0])
                        runtime += [time.time() - startTime]
                  print("WHILE: Calculation using {} threads took {:.4f} +- {:.4f}".format(op[1], np.mean(runtime), np.std(runtime) ))
                
            for op in paraOpsMap:
                  runtime = []
                  sess.run(op[0]) #warm up
                  for i in range(runs):
                        startTime = time.time()
                        sess.run(op[0])
                        runtime += [time.time() - startTime]
                  print("MAP: Calculation using {} threads took {:.4f} +- {:.4f}".format(op[1], np.mean(runtime), np.std(runtime) ))
```
 Example Output:

```
Calculation using 1 threads took 0.0044 +- 0.0004
Calculation using 10 threads took 0.0026 +- 0.0006
Calculation using 50 threads took 0.0018 +- 0.0000
Calculation using 100 threads took 0.0014 +- 0.0001
WHILE: Calculation using 1 threads took 0.0063 +- 0.0001
WHILE: Calculation using 10 threads took 0.0063 +- 0.0007
WHILE: Calculation using 50 threads took 0.0063 +- 0.0001
WHILE: Calculation using 100 threads took 0.0062 +- 0.0004
MAP: Calculation using 1 threads took 0.0072 +- 0.0004
MAP: Calculation using 10 threads took 0.0035 +- 0.0003
MAP: Calculation using 50 threads took 0.0035 +- 0.0002
MAP: Calculation using 100 threads took 0.0035 +- 0.0002

```

 

### System information
Ubuntu, Tensorflow installed using pip, version 1.3
tested on two different systems

**Comment by reedwm** [comment]
@skye can you take a look?

**Comment by kelvinxu** [comment]
Any update on this?

**Comment by skye** [comment]
Looking at calRow_while, it appears you're using control dependencies to serialize the while loop iterations. This is probably why you're not seeing speedups from increasing the parallel_iterations argument (since it can't actually run anything in parallel).

**Comment by georgh** [comment]
This could be the case, but how would you perform it without the dependencies?
I thought that is the default way to use while_loop?

**Comment by skye** [comment]
It depends on the body of your while loop. Sometimes there are independent calculations that can be run in parallel (e.g. if you have many different loop variables and they don't all depend on each other). Also, I suspect you're seeing the manual calculation run faster than the while loop because the body computation is small compared to the overhead of running the while loop. A more complicated or expensive body would probably see less of a difference.

**Comment by georgh** [comment]
The above calculation is only an example, the body I use takes quiet some time.
But for the above body - how would you write it down to achieve parallel execution with while_loop?

**Comment by skye** [comment]
Remove the control dependency block? It depends on what you're trying to achieve... if the variable assigns can happen in any order, you don't need the control dependencies and you can run in parallel. If the assigns must happen in order, then you can't run them in parallel and this is enforced via control deps.

Note that map_fn calls tf.while_loop() under the hood, but doesn't add any control dependencies. So that's an example of parallel execution using while loop.

**Comment by georgh** [comment]
Ok now I got what you are suggesting. 
The following does **not** work:
```
            def body(ic):
                  updateKernel = calEntry(KernelVariable, ic)
                        return ic + 1
```
This will never execute calEntry.

But if you use 
```
            i = tf.constant(0)
            m0 = tf.constant(0)
            condition = lambda i, m: tf.less(i, N)
            def body(ic, dontcare):
                  updateKernel = calEntry(KernelVariable, ic)
                  return ic + 1, updateKernel
            return tf.while_loop(condition, body, [i,m0], back_prop=False, parallel_iterations=parallelNum)
```
It works and is as fast as using map or foldl, foldr
Thank you a lot for your help, even though it does not change much because now my code already uses map and fold everywhere. 

You can still gain some speed if you use 
```
            for ic in range(0, N, parallelNum):
                  with tf.control_dependencies(ops):
                        for j in range(parallelNum):
                              if ic + j < N:
                                    ops += [calEntry(KernelVariable, ic+j)]
```
but this may result in a very slow graph creation. 
@skye is there any plan to add a parallel while_loop with lesser overhead?

**Comment by skye** [comment]
You're welcome! Using the functional ops will probably make your program easier to understand than using while_loop directly (like I said, these functions call while_loop under the hood).

There are no plans to change while_loop right now, no. I suggest benchmarking your actual application, as I expect you'll find the overhead to be a smaller percentage of the total runtime than in this microbenchmark.

I'm gonna close this issue now.

**Comment by kelvinxu** [comment]
Hi @skye  I encountered something similar when running tf.map_fn where the function was a neural network that did a bunch of gradient steps. The inputs were just slices of a minibatch. 

Basically this: https://github.com/cbfinn/maml/blob/master/maml.py#L78

I noticed that on CPU, there was parallelization, but on GPU no. Might there be a reason for this?

It seems like the post here was not the same issue.
